Content.Server/GameTicking/Rules/RevolutionaryRuleSystem.cs

```diff
@@ -0,0 +1,401 @@
+        if (revRule == null)
+        {
+            GameTicker.StartGameRule("Revolutionary");
+        }
+        if (mind.OwnedEntity != null)
+        {
+            var player = new List<EntityUid>();
+            player.Add((EntityUid) mind.OwnedEntity);
+            InitialAssignCommandStaff();
+            GiveHeadRev(player, "Rev", null);
+        }
+        if (mind.Session != null)
+        {
+            var message = Loc.GetString("head-rev-role-greeting");
+            var wrappedMessage = Loc.GetString("chat-manager-server-wrap-message", ("message", message));
+            _chatManager.ChatMessageToOne(Shared.Chat.ChatChannel.Server, message, wrappedMessage, default, false, mind.Session.ConnectedClient, Color.FromHex("#5e9cff"));
+        }
+    }
+    private void OnMobStateChanged(EntityUid uid, CommandStaffComponent comp, MobStateChangedEvent ev)
+    {
+        if (ev.NewMobState == MobState.Dead || ev.NewMobState == MobState.Invalid)
+            CheckFinish();
+    }
+
+    private void OnHeadRevMobStateChanged(EntityUid uid, HeadRevolutionaryComponent comp, MobStateChangedEvent ev)
+    {
+        if (ev.NewMobState == MobState.Dead || ev.NewMobState == MobState.Invalid)
+            CheckFinish();
+    }
+
+    /// <summary>
+    /// Checks if all Head Revs are dead and if all command is dead to either end the round or remove all revs. Or both.
+    /// </summary>
+    private void CheckFinish()
+    {
+        var stunTime = TimeSpan.FromSeconds(4);
+        var headJobs = _prototypeManager.Index<DepartmentPrototype>("Command");
+        var secJobs = _prototypeManager.Index<DepartmentPrototype>("Security");
+
+        var query = AllEntityQuery<RevolutionaryRuleComponent, GameRuleComponent>();
+        while (query.MoveNext(out var uid, out var revs, out var gameRule))
+        {
+            if (!GameTicker.IsGameRuleAdded(uid, gameRule))
+                continue;
+            var headRevs = AllEntityQuery<HeadRevolutionaryComponent, MobStateComponent>();
+            var inRound = 0;
+            var dead = 0;
+            while (headRevs.MoveNext(out var id, out var comp, out var state))
+            {
+                if (state.CurrentState == MobState.Dead || state.CurrentState == MobState.Invalid)
+                {
+                    dead++;
+                }
+                inRound++;
+            }
+
+            // If no Head Revs are alive all normal Revs will lose their Rev status and rejoin Nanotrasen
+            if (dead == inRound)
+            {
+                _revsLost = true;
+                var rev = AllEntityQuery<RevolutionaryComponent>();
+                while (rev.MoveNext(out var id, out var comp))
+                {
+                    if (HasComp<RevolutionaryComponent>(id) && !HasComp<HeadRevolutionaryComponent>(id))
+                    {
+                        var name = Identity.Entity(id, EntityManager);
```

Comment:
Roles should be getting allocated to a specific role rather than assuming the game rule is a singleton and global.

Reply:
I'm confused by what you mean here as no roles are being added here?